### PR TITLE
Fix support for multiple pvc for tikv

### PR DESCRIPTION
**Pull request by dragonly**
### What problem does this PR solve?

ref: #3802

### What is changed and how does it work?

Change tikv related code to support multiple pvc from pod spec, instead of constructing a single pvc name from tc name and ordinal.

Also refactored related util functions for brevity.

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [x] Manual test 
    1. deploy a tc
    ```yaml
    apiVersion: pingcap.com/v1alpha1
    kind: TidbCluster
    metadata:
        name: basic
    spec:
        version: v4.0.10
        timezone: UTC
        pvReclaimPolicy: Delete
        configUpdateStrategy: RollingUpdate
        discovery: {}
        pd:
            baseImage: pingcap/pd
            replicas: 3
            requests:
                storage: "1Gi"
            config: {}
        tikv:
            baseImage: pingcap/tikv
            replicas: 4
            requests:
                storage: "1Gi"
            storageVolumes:
                - name: additional
                storageSize: "1Gi"
                mountPath: /mnt
            config: {}
        tidb:
            baseImage: pingcap/tidb
            replicas: 1
            config: {}
    ```
    2. check pvc with `kubectl get pvc`
    ```
    NAME                           STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
    pd-basic-pd-0                  Bound    pvc-8d98e8d5-93de-432a-9d19-b2b3e8826282   1Gi        RWO            standard       103m
    pd-basic-pd-1                  Bound    pvc-69458bd2-290b-49a7-aff8-029408498ac4   1Gi        RWO            standard       103m
    pd-basic-pd-2                  Bound    pvc-a571b440-b7dd-4f1e-bd99-327e46b2b843   1Gi        RWO            standard       103m
    tikv-additional-basic-tikv-0   Bound    pvc-eb8ec28f-b78e-4f83-89b1-e1e1e172f127   1Gi        RWO            standard       102m
    tikv-additional-basic-tikv-1   Bound    pvc-ed541ee0-5bee-40bb-ab94-34eaa65478df   1Gi        RWO            standard       102m
    tikv-additional-basic-tikv-2   Bound    pvc-45eab6ed-16e0-42fb-9f96-8f9e9d2db97e   1Gi        RWO            standard       102m
    tikv-additional-basic-tikv-3   Bound    pvc-70699ee8-d350-4208-99ba-68d837e99eeb   1Gi        RWO            standard       102m
    tikv-basic-tikv-0              Bound    pvc-8b3a8886-1804-4b68-b337-1304758ba3f8   1Gi        RWO            standard       102m
    tikv-basic-tikv-1              Bound    pvc-6a3fd2cd-99a4-4b2b-b5c6-1a43460fed1a   1Gi        RWO            standard       102m
    tikv-basic-tikv-2              Bound    pvc-8046ef23-b6e5-44e1-8813-cb0fe0913985   1Gi        RWO            standard       102m
    tikv-basic-tikv-3              Bound    pvc-9013e429-6037-4984-83f3-f2310f71b0a6   1Gi        RWO            standard       102m
    ```
    3. scale in tikv to 3 replicas, check pvc again, returns same results as in step 2, because PVCs are retained by default.
    ```
    NAME                           STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
    pd-basic-pd-0                  Bound    pvc-8d98e8d5-93de-432a-9d19-b2b3e8826282   1Gi        RWO            standard       103m
    pd-basic-pd-1                  Bound    pvc-69458bd2-290b-49a7-aff8-029408498ac4   1Gi        RWO            standard       103m
    pd-basic-pd-2                  Bound    pvc-a571b440-b7dd-4f1e-bd99-327e46b2b843   1Gi        RWO            standard       103m
    tikv-additional-basic-tikv-0   Bound    pvc-eb8ec28f-b78e-4f83-89b1-e1e1e172f127   1Gi        RWO            standard       102m
    tikv-additional-basic-tikv-1   Bound    pvc-ed541ee0-5bee-40bb-ab94-34eaa65478df   1Gi        RWO            standard       102m
    tikv-additional-basic-tikv-2   Bound    pvc-45eab6ed-16e0-42fb-9f96-8f9e9d2db97e   1Gi        RWO            standard       102m
    tikv-additional-basic-tikv-3   Bound    pvc-70699ee8-d350-4208-99ba-68d837e99eeb   1Gi        RWO            standard       102m
    tikv-basic-tikv-0              Bound    pvc-8b3a8886-1804-4b68-b337-1304758ba3f8   1Gi        RWO            standard       102m
    tikv-basic-tikv-1              Bound    pvc-6a3fd2cd-99a4-4b2b-b5c6-1a43460fed1a   1Gi        RWO            standard       102m
    tikv-basic-tikv-2              Bound    pvc-8046ef23-b6e5-44e1-8813-cb0fe0913985   1Gi        RWO            standard       102m
    tikv-basic-tikv-3              Bound    pvc-9013e429-6037-4984-83f3-f2310f71b0a6   1Gi        RWO            standard       102m
    ```
    4. scale out tikv to 4 replicas, check pvc again, found that `tikv-basic-tikv-3` and `tikv-additional-basic-tikv-3` has changed, because the old PVC is deleted and new one is created with the same name. we can tell from the different `VOLUME` name.
    ```
    NAME                           STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
    pd-basic-pd-0                  Bound    pvc-8d98e8d5-93de-432a-9d19-b2b3e8826282   1Gi        RWO            standard       104m
    pd-basic-pd-1                  Bound    pvc-69458bd2-290b-49a7-aff8-029408498ac4   1Gi        RWO            standard       104m
    pd-basic-pd-2                  Bound    pvc-a571b440-b7dd-4f1e-bd99-327e46b2b843   1Gi        RWO            standard       104m
    tikv-additional-basic-tikv-0   Bound    pvc-eb8ec28f-b78e-4f83-89b1-e1e1e172f127   1Gi        RWO            standard       103m
    tikv-additional-basic-tikv-1   Bound    pvc-ed541ee0-5bee-40bb-ab94-34eaa65478df   1Gi        RWO            standard       103m
    tikv-additional-basic-tikv-2   Bound    pvc-45eab6ed-16e0-42fb-9f96-8f9e9d2db97e   1Gi        RWO            standard       103m
    tikv-additional-basic-tikv-3   Bound    pvc-2911848a-03f9-4406-8495-502f418d2fd6   1Gi        RWO            standard       6s
    tikv-basic-tikv-0              Bound    pvc-8b3a8886-1804-4b68-b337-1304758ba3f8   1Gi        RWO            standard       103m
    tikv-basic-tikv-1              Bound    pvc-6a3fd2cd-99a4-4b2b-b5c6-1a43460fed1a   1Gi        RWO            standard       103m
    tikv-basic-tikv-2              Bound    pvc-8046ef23-b6e5-44e1-8813-cb0fe0913985   1Gi        RWO            standard       103m
    tikv-basic-tikv-3              Bound    pvc-4fb8a2b2-8b07-4603-95ef-c71df247cc84   1Gi        RWO            standard       6s
    ```
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Fix support for multiple PVC for TiKV.
```

### CODE REVIEW

File: pkg/manager/member/tikv_scaler.go [hunk 1]
```diff
@@ -115,6 +114,7 @@ func (s *tikvScaler) ScaleIn(meta metav1.Object, oldSet *apps.StatefulSet, newSe
 		return nil
 	}
 
+	// call pd to delete store of the will-be-scaled-in tikv pod
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler.go:117]:
```suggestion
	// call PD API to delete the store of the TiKV Pod to be scaled in
```

File: pkg/manager/member/tikv_scaler.go [hunk 2]
```diff
@@ -132,8 +132,10 @@ func (s *tikvScaler) ScaleIn(meta metav1.Object, oldSet *apps.StatefulSet, newSe
 			return controller.RequeueErrorf("TiKV %s/%s store %d is still in cluster, state: %s", ns, podName, id, state)
 		}
 	}
-	for id, store := range tc.Status.TiKV.TombstoneStores {
-		if store.PodName == podName && pod.Labels[label.StoreIDLabelKey] == id {
+
+	// annotate tikv pods with tomestone store
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler.go:136]:
```suggestion
	// If the store state turns to Tombstone, add defer deleting annotation to the PVCs of the Pod
```

File: pkg/manager/member/tikv_scaler.go [hunk 3]
```diff
@@ -142,24 +144,24 @@ func (s *tikvScaler) ScaleIn(meta metav1.Object, oldSet *apps.StatefulSet, newSe
 			// TODO: double check if store is really not in Up/Offline/Down state
 			klog.Infof("TiKV %s/%s store %d becomes tombstone", ns, podName, id)
 
-			pvcName := ordinalPVCName(v1alpha1.TiKVMemberType, setName, ordinal)
-			pvc, err := s.deps.PVCLister.PersistentVolumeClaims(ns).Get(pvcName)
+			pvcs, err := util.ResolvePVCFromPod(pod, s.deps.PVCLister)
+			klog.V(4).Infof("ResolvePVCFromPod: %+v", pvcs)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler.go:148]: Is it necessary to print the full object of each PVC?

Comment (DanielZhangQD) [reply pkg/manager/member/tikv_scaler.go:148]: And move this after L151?

File: pkg/manager/member/tikv_scaler.go [hunk 4]
```diff
@@ -142,24 +144,24 @@ func (s *tikvScaler) ScaleIn(meta metav1.Object, oldSet *apps.StatefulSet, newSe
...
+					pvc.Annotations = map[string]string{}
+				}
+				now := time.Now().Format(time.RFC3339)
+				pvc.Annotations[label.AnnPVCDeferDeleting] = now
+				_, err = s.deps.PVCControl.UpdatePVC(tc, pvc)
+				if err != nil {
+					klog.Errorf("tikv scale in: failed to set pvc %s/%s annotation: %s to %s", ns, pvc.Name, label.AnnPVCDeferDeleting, now)
+					return err
+				}
+				klog.Infof("tikv scale in: set pvc %s/%s annotation: %s to %s", ns, pvc.Name, label.AnnPVCDeferDeleting, now)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler.go:163]: Define a function for this that can be used by other functions?

File: pkg/manager/member/tikv_scaler.go [hunk 5]
```diff
@@ -171,17 +173,16 @@ func (s *tikvScaler) ScaleIn(meta metav1.Object, oldSet *apps.StatefulSet, newSe
...
+	// When store not found in TidbCluster status, there are two possible situations:
+	// 1. TiKV has already joined cluster but status not synced yet.
+	//    In this situation return error to wait for another round for safety.
+	// 2. TiKV pod is not ready, such as in pending state.
 	//    In this situation we should delete this TiKV pod immediately to avoid blocking the subsequent operations.
 	if !podutil.IsPodReady(pod) {
-		pvcName := ordinalPVCName(v1alpha1.TiKVMemberType, setName, ordinal)
-		pvc, err := s.deps.PVCLister.PersistentVolumeClaims(ns).Get(pvcName)
+		pvcs, err := util.ResolvePVCFromPod(pod, s.deps.PVCLister)
+		klog.V(4).Infof("ResolvePVCFromPod: %+v", pvcs)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler.go:183]: ditto.

File: pkg/manager/member/tikv_scaler.go [hunk 6]
```diff
@@ -171,17 +173,16 @@ func (s *tikvScaler) ScaleIn(meta metav1.Object, oldSet *apps.StatefulSet, newSe
...
 		if err != nil {
-			return fmt.Errorf("tikvScaler.ScaleIn: failed to get pvc %s for cluster %s/%s, error: %s", pvcName, ns, tcName, err)
+			return fmt.Errorf("tikvScaler.ScaleIn: failed to get pvcs for pod %s/%s in tc %s/%s, error: %s", ns, pod.Name, ns, tcName, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler.go:185]: Move this block after L200?

File: pkg/manager/member/tikv_scaler.go [hunk 7]
```diff
@@ -194,23 +195,23 @@ func (s *tikvScaler) ScaleIn(meta metav1.Object, oldSet *apps.StatefulSet, newSe
...
+			}
+			now := time.Now().Format(time.RFC3339)
+			pvc.Annotations[label.AnnPVCDeferDeleting] = now
+			_, err = s.deps.PVCControl.UpdatePVC(tc, pvc)
+			if err != nil {
+				klog.Errorf("pod %s not ready, tikv scale in: failed to set pvc %s/%s annotation: %s to %s", podName, ns, pvc.Name, label.AnnPVCDeferDeleting, now)
+				return err
+			}
+			klog.Infof("pod %s not ready, tikv scale in: set pvc %s/%s annotation: %s to %s", podName, ns, pvc.Name, label.AnnPVCDeferDeleting, now)
+			setReplicasAndDeleteSlots(newSet, replicas, deleteSlots)
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler.go:213]: Why move this to the loop?

File: pkg/util/util.go [hunk 8]
```diff
@@ -430,13 +431,14 @@ func ResolvePVCFromPod(pod *corev1.Pod, pvcLister corelisterv1.PersistentVolumeC
...
 				klog.Errorf("Get PVC %s/%s error: %v", pod.Namespace, pvcName, err)
-				continue
+				return nil, err
 			}
 			pvcs = append(pvcs, pvc)
 		}
 	}
 	if len(pvcs) == 0 {
-		return nil, errors.NewNotFound(corev1.Resource("pvc"), pod.Name)
+		klog.Errorf("no pvc found for pod %s", pod.Name)
```
Comment (DanielZhangQD) [inline pkg/util/util.go:440]: Add namespace.

File: pkg/util/util.go [hunk 9]
```diff
@@ -430,13 +431,14 @@ func ResolvePVCFromPod(pod *corev1.Pod, pvcLister corelisterv1.PersistentVolumeC
...
+		return nil, errors.NewNotFound(corev1.Resource("pvc"), "")
```
Comment (DanielZhangQD) [inline pkg/util/util.go:441]: The error message is not clear enough, what about just return new error with the log message in the last line.

File: pkg/manager/member/tikv_scaler.go [hunk 10]
```diff
@@ -142,24 +144,24 @@ func (s *tikvScaler) ScaleIn(meta metav1.Object, oldSet *apps.StatefulSet, newSe
...
+				return fmt.Errorf("tikvScaler.ScaleIn: failed to get pvcs for pod %s/%s in tc %s/%s, error: %s", ns, pod.Name, ns, tcName, err)
 			}
-			now := time.Now().Format(time.RFC3339)
-			pvc.Annotations[label.AnnPVCDeferDeleting] = now
-			_, err = s.deps.PVCControl.UpdatePVC(tc, pvc)
-			if err != nil {
-				klog.Errorf("tikv scale in: failed to set pvc %s/%s annotation: %s to %s",
-					ns, pvcName, label.AnnPVCDeferDeleting, now)
-				return err
+			for _, pvc := range pvcs {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler.go:152]: Suggest updating the e2e cases to cover the code change in this file.

File: pkg/util/util.go (hunk 9 above)
Comment (dragonly) [reply pkg/util/util.go:441]: maybe the caller will check error type with `errors.IsNotFound()` and skip a `NotFound` error?

File: pkg/manager/member/tikv_scaler.go (hunk 7 above)
Comment (dragonly) [reply pkg/manager/member/tikv_scaler.go:213]: by mistake, fixed in next commits

File: pkg/util/util.go (hunk 9 above)
Comment (DanielZhangQD) [reply pkg/util/util.go:441]: The current callers do not check the error type now and if some do check later, we can define a new NotFound error with a clearer message, the current one will just report no PVC found, without any useful info, e.g., which PVC, for which Pod in which ns, right?

File: pkg/manager/member/utils.go [hunk 11]
```diff
@@ -501,3 +502,18 @@ func CreateOrUpdateService(serviceLister corelisters.ServiceLister, serviceContr
 	}
 	return nil
 }
+
+// AnnotatePVCDeferDeletingLabel set the label
+func AnnotatePVCDeferDeletingLabel(tc *v1alpha1.TidbCluster, pvc *corev1.PersistentVolumeClaim, pvcControl controller.PVCControlInterface) error {
```
Comment (DanielZhangQD) [inline pkg/manager/member/utils.go:507]:
```suggestion
func addDeferDeletingAnnoToPVC(tc *v1alpha1.TidbCluster, pvc *corev1.PersistentVolumeClaim, pvcControl controller.PVCControlInterface) error {
```

File: pkg/manager/member/tikv_scaler.go (hunk 10 above)
Comment (dragonly) [reply pkg/manager/member/tikv_scaler.go:151]: will add e2e case in another PR

File: pkg/util/util.go (hunk 9 above)
Comment (dragonly) [reply pkg/util/util.go]: PD has [ut case](https://github.com/pingcap/tidb-operator/blob/master/pkg/manager/member/pd_failover_test.go#L323) that asserts no pvc found but deleting pod succeeded.
Fixed in https://github.com/pingcap/tidb-operator/pull/3820/files#diff-0fa449830b5dc2b274b7c95a58b5071edc23f1e44af84a99c1fa9e600767829fR440

File: pkg/manager/member/tikv_scaler.go (hunk 10 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tikv_scaler.go:151]: ok

**Comment by DanielZhangQD** [comment]
@dragonly Please add manual test results in the PR description.

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by handlerww — APPROVED** [review]
LGTM

File: pkg/manager/member/tikv_scaler_test.go [hunk 12]
```diff
@@ -200,6 +199,13 @@ func TestTiKVScalerScaleIn(t *testing.T) {
 		if test.hasPVC {
 			pvc := newScaleInPVCForStatefulSet(oldSet, v1alpha1.TiKVMemberType, tc.Name)
 			pvcIndexer.Add(pvc)
+			pod.Spec.Volumes = append(pod.Spec.Volumes, corev1.Volume{
+				VolumeSource: corev1.VolumeSource{
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_scaler_test.go:203]: Should we cover the multiple PVC case?

File: pkg/util/util.go [hunk 13]
```diff
@@ -430,13 +430,13 @@ func ResolvePVCFromPod(pod *corev1.Pod, pvcLister corelisterv1.PersistentVolumeC
...
 				klog.Errorf("Get PVC %s/%s error: %v", pod.Namespace, pvcName, err)
-				continue
+				return nil, err
 			}
 			pvcs = append(pvcs, pvc)
 		}
 	}
 	if len(pvcs) == 0 {
-		return nil, errors.NewNotFound(corev1.Resource("pvc"), pod.Name)
+		return nil, fmt.Errorf("no pvc found for pod %s/%s", pod.Namespace, pod.Name)
```
Comment (DanielZhangQD) [inline pkg/util/util.go:439]: Why return the error here?
For this function, the Pod may include PVC or not, both cases are possible behavior, and the caller should decide whether `len(pvcs) == 0` is expected or not and handle this case.

File: pkg/manager/member/tikv_scaler_test.go (hunk 12 above)
Comment (dragonly) [reply pkg/manager/member/tikv_scaler_test.go]: added

File: pkg/util/util.go (hunk 13 above)
Comment (dragonly) [reply pkg/util/util.go]: fixed.

**Pull request merged**